// lilybird/dist/enums/application.js
export var ApplicationFlags;
(function (ApplicationFlags) {
    ApplicationFlags[ApplicationFlags["APPLICATION_AUTO_MODERATION_RULE_CREATE_BADGE"] = 64] = "APPLICATION_AUTO_MODERATION_RULE_CREATE_BADGE";
    ApplicationFlags[ApplicationFlags["GATEWAY_PRESENCE"] = 4096] = "GATEWAY_PRESENCE";
    ApplicationFlags[ApplicationFlags["GATEWAY_PRESENCE_LIMITED"] = 8192] = "GATEWAY_PRESENCE_LIMITED";
    ApplicationFlags[ApplicationFlags["GATEWAY_GUILD_MEMBERS"] = 16384] = "GATEWAY_GUILD_MEMBERS";
    ApplicationFlags[ApplicationFlags["GATEWAY_GUILD_MEMBERS_LIMITED"] = 32768] = "GATEWAY_GUILD_MEMBERS_LIMITED";
    ApplicationFlags[ApplicationFlags["VERIFICATION_PENDING_GUILD_LIMIT"] = 65536] = "VERIFICATION_PENDING_GUILD_LIMIT";
    ApplicationFlags[ApplicationFlags["EMBEDDED"] = 131072] = "EMBEDDED";
    ApplicationFlags[ApplicationFlags["GATEWAY_MESSAGE_CONTENT"] = 262144] = "GATEWAY_MESSAGE_CONTENT";
    ApplicationFlags[ApplicationFlags["GATEWAY_MESSAGE_CONTENT_LIMITED"] = 524288] = "GATEWAY_MESSAGE_CONTENT_LIMITED";
    ApplicationFlags[ApplicationFlags["APPLICATION_COMMAND_BADGE"] = 8388608] = "APPLICATION_COMMAND_BADGE";
})(ApplicationFlags || (ApplicationFlags = {}));
export var ApplicationRoleConnectionMetadataType;
(function (ApplicationRoleConnectionMetadataType) {
    ApplicationRoleConnectionMetadataType[ApplicationRoleConnectionMetadataType["INTEGER_LESS_THAN_OR_EQUAL"] = 1] = "INTEGER_LESS_THAN_OR_EQUAL";
    ApplicationRoleConnectionMetadataType[ApplicationRoleConnectionMetadataType["INTEGER_GREATER_THAN_OR_EQUAL"] = 2] = "INTEGER_GREATER_THAN_OR_EQUAL";
    ApplicationRoleConnectionMetadataType[ApplicationRoleConnectionMetadataType["INTEGER_EQUAL"] = 3] = "INTEGER_EQUAL";
    ApplicationRoleConnectionMetadataType[ApplicationRoleConnectionMetadataType["INTEGER_NOT_EQUAL"] = 4] = "INTEGER_NOT_EQUAL";
    ApplicationRoleConnectionMetadataType[ApplicationRoleConnectionMetadataType["DATETIME_LESS_THAN_OR_EQUAL"] = 5] = "DATETIME_LESS_THAN_OR_EQUAL";
    ApplicationRoleConnectionMetadataType[ApplicationRoleConnectionMetadataType["DATETIME_GREATER_THAN_OR_EQUAL"] = 6] = "DATETIME_GREATER_THAN_OR_EQUAL";
    ApplicationRoleConnectionMetadataType[ApplicationRoleConnectionMetadataType["BOOLEAN_EQUAL"] = 7] = "BOOLEAN_EQUAL";
    ApplicationRoleConnectionMetadataType[ApplicationRoleConnectionMetadataType["BOOLEAN_NOT_EQUAL"] = 8] = "BOOLEAN_NOT_EQUAL";
})(ApplicationRoleConnectionMetadataType || (ApplicationRoleConnectionMetadataType = {}));
export var ApplicationIntegrationType;
(function (ApplicationIntegrationType) {
    ApplicationIntegrationType[ApplicationIntegrationType["GUILD_INSTALL"] = 0] = "GUILD_INSTALL";
    ApplicationIntegrationType[ApplicationIntegrationType["USER_INSTALL"] = 1] = "USER_INSTALL";
})(ApplicationIntegrationType || (ApplicationIntegrationType = {}));
