// lilybird/dist/enums/auto-moderation.js
export var AutoModerationEventType;
(function (AutoModerationEventType) {
    AutoModerationEventType[AutoModerationEventType["MESSAGE_SEND"] = 1] = "MESSAGE_SEND";
})(AutoModerationEventType || (AutoModerationEventType = {}));
export var AutoModerationTriggerType;
(function (AutoModerationTriggerType) {
    AutoModerationTriggerType[AutoModerationTriggerType["KEYWORD"] = 1] = "KEYWORD";
    AutoModerationTriggerType[AutoModerationTriggerType["SPAM"] = 3] = "SPAM";
    AutoModerationTriggerType[AutoModerationTriggerType["KEYWORD_PRESET"] = 4] = "KEYWORD_PRESET";
    AutoModerationTriggerType[AutoModerationTriggerType["MENTION_SPAM"] = 5] = "MENTION_SPAM";
})(AutoModerationTriggerType || (AutoModerationTriggerType = {}));
export var AutoModerationKeywordPresetType;
(function (AutoModerationKeywordPresetType) {
    AutoModerationKeywordPresetType[AutoModerationKeywordPresetType["PROFANITY"] = 1] = "PROFANITY";
    AutoModerationKeywordPresetType[AutoModerationKeywordPresetType["SEXUAL_CONTENT"] = 2] = "SEXUAL_CONTENT";
    AutoModerationKeywordPresetType[AutoModerationKeywordPresetType["SLURS"] = 3] = "SLURS";
})(AutoModerationKeywordPresetType || (AutoModerationKeywordPresetType = {}));
export var AutoModerationActionType;
(function (AutoModerationActionType) {
    AutoModerationActionType[AutoModerationActionType["BLOCK_MESSAGE"] = 1] = "BLOCK_MESSAGE";
    AutoModerationActionType[AutoModerationActionType["SEND_ALERT_MESSAGE"] = 2] = "SEND_ALERT_MESSAGE";
    AutoModerationActionType[AutoModerationActionType["TIMEOUT"] = 3] = "TIMEOUT";
})(AutoModerationActionType || (AutoModerationActionType = {}));
