// lilybird/dist/enums/channel.js
export var ChannelType;
(function (ChannelType) {
    ChannelType[ChannelType["GUILD_TEXT"] = 0] = "GUILD_TEXT";
    ChannelType[ChannelType["DM"] = 1] = "DM";
    ChannelType[ChannelType["GUILD_VOICE"] = 2] = "GUILD_VOICE";
    ChannelType[ChannelType["GROUP_DM"] = 3] = "GROUP_DM";
    ChannelType[ChannelType["GUILD_CATEGORY"] = 4] = "GUILD_CATEGORY";
    ChannelType[ChannelType["GUILD_ANNOUNCEMENT"] = 5] = "GUILD_ANNOUNCEMENT";
    ChannelType[ChannelType["ANNOUNCEMENT_THREAD"] = 10] = "ANNOUNCEMENT_THREAD";
    ChannelType[ChannelType["PUBLIC_THREAD"] = 11] = "PUBLIC_THREAD";
    ChannelType[ChannelType["PRIVATE_THREAD"] = 12] = "PRIVATE_THREAD";
    ChannelType[ChannelType["GUILD_STAGE_VOICE"] = 13] = "GUILD_STAGE_VOICE";
    ChannelType[ChannelType["GUILD_DIRECTORY"] = 14] = "GUILD_DIRECTORY";
    ChannelType[ChannelType["GUILD_FORUM"] = 15] = "GUILD_FORUM";
    ChannelType[ChannelType["GUILD_MEDIA"] = 16] = "GUILD_MEDIA";
})(ChannelType || (ChannelType = {}));
export var OverwriteType;
(function (OverwriteType) {
    OverwriteType[OverwriteType["ROLE"] = 0] = "ROLE";
    OverwriteType[OverwriteType["MEMBER"] = 1] = "MEMBER";
})(OverwriteType || (OverwriteType = {}));
export var VideoQualityMode;
(function (VideoQualityMode) {
    VideoQualityMode[VideoQualityMode["AUTO"] = 1] = "AUTO";
    VideoQualityMode[VideoQualityMode["FULL"] = 2] = "FULL";
})(VideoQualityMode || (VideoQualityMode = {}));
export var ChannelFlags;
(function (ChannelFlags) {
    ChannelFlags[ChannelFlags["PINNED"] = 2] = "PINNED";
    ChannelFlags[ChannelFlags["REQUIRE_TAG"] = 16] = "REQUIRE_TAG";
    ChannelFlags[ChannelFlags["HIDE_MEDIA_DOWNLOAD_OPTIONS"] = 32768] = "HIDE_MEDIA_DOWNLOAD_OPTIONS";
})(ChannelFlags || (ChannelFlags = {}));
export var SortOrderType;
(function (SortOrderType) {
    SortOrderType[SortOrderType["LATEST_ACTIVITY"] = 0] = "LATEST_ACTIVITY";
    SortOrderType[SortOrderType["CREATION_DATE"] = 1] = "CREATION_DATE";
})(SortOrderType || (SortOrderType = {}));
export var ForumLayoutType;
(function (ForumLayoutType) {
    ForumLayoutType[ForumLayoutType["NOT_SET"] = 0] = "NOT_SET";
    ForumLayoutType[ForumLayoutType["LIST_VIEW"] = 1] = "LIST_VIEW";
    ForumLayoutType[ForumLayoutType["GALLERY_VIEW"] = 2] = "GALLERY_VIEW";
})(ForumLayoutType || (ForumLayoutType = {}));
export var AllowedMentionType;
(function (AllowedMentionType) {
    AllowedMentionType["RoleMentions"] = "roles";
    AllowedMentionType["UserMentions"] = "users";
    AllowedMentionType["EveryoneMentions"] = "everyone";
})(AllowedMentionType || (AllowedMentionType = {}));
