// lilybird/dist/enums/client.js
export var TransformerReturnType;
(function (TransformerReturnType) {
    TransformerReturnType[TransformerReturnType["SINGLE"] = 0] = "SINGLE";
    TransformerReturnType[TransformerReturnType["MULTIPLE"] = 1] = "MULTIPLE";
})(TransformerReturnType || (TransformerReturnType = {}));
export var DebugIdentifier;
(function (DebugIdentifier) {
    DebugIdentifier[DebugIdentifier["WSMessage"] = 0] = "WSMessage";
    DebugIdentifier[DebugIdentifier["Heartbeat"] = 1] = "Heartbeat";
    DebugIdentifier[DebugIdentifier["ACK"] = 2] = "ACK";
    DebugIdentifier[DebugIdentifier["HeartbeatRequest"] = 3] = "HeartbeatRequest";
    DebugIdentifier[DebugIdentifier["Identify"] = 4] = "Identify";
    DebugIdentifier[DebugIdentifier["Resume"] = 5] = "Resume";
    DebugIdentifier[DebugIdentifier["InvalidSession"] = 6] = "InvalidSession";
    DebugIdentifier[DebugIdentifier["Reconnect"] = 7] = "Reconnect";
    DebugIdentifier[DebugIdentifier["WSError"] = 8] = "WSError";
    DebugIdentifier[DebugIdentifier["MissingACK"] = 9] = "MissingACK";
    DebugIdentifier[DebugIdentifier["ZombieConnection"] = 10] = "ZombieConnection";
    DebugIdentifier[DebugIdentifier["AttemptingResume"] = 11] = "AttemptingResume";
    DebugIdentifier[DebugIdentifier["CloseCode"] = 12] = "CloseCode";
    DebugIdentifier[DebugIdentifier["CompiledListeners"] = 13] = "CompiledListeners";
    DebugIdentifier[DebugIdentifier["RESTCall"] = 14] = "RESTCall";
    DebugIdentifier[DebugIdentifier["RESTError"] = 15] = "RESTError";
})(DebugIdentifier || (DebugIdentifier = {}));
