// lilybird/dist/enums/gateway.js
export var GatewayEvent;
(function (GatewayEvent) {
    GatewayEvent["Ready"] = "READY";
    GatewayEvent["Resumed"] = "RESUMED";
    GatewayEvent["ApplicationCommandPermissionsUpdate"] = "APPLICATION_COMMAND_PERMISSIONS_UPDATE";
    GatewayEvent["AutoModerationRuleCreate"] = "AUTO_MODERATION_RULE_CREATE";
    GatewayEvent["AutoModerationRuleUpdate"] = "AUTO_MODERATION_RULE_UPDATE";
    GatewayEvent["AutoModerationRuleDelete"] = "AUTO_MODERATION_RULE_DELETE";
    GatewayEvent["AutoModerationActionExecution"] = "AUTO_MODERATION_ACTION_EXECUTION";
    GatewayEvent["ChannelCreate"] = "CHANNEL_CREATE";
    GatewayEvent["ChannelUpdate"] = "CHANNEL_UPDATE";
    GatewayEvent["ChannelDelete"] = "CHANNEL_DELETE";
    GatewayEvent["ChannelPinsUpdate"] = "CHANNEL_PINS_UPDATE";
    GatewayEvent["ThreadCreate"] = "THREAD_CREATE";
    GatewayEvent["ThreadUpdate"] = "THREAD_UPDATE";
    GatewayEvent["ThreadDelete"] = "THREAD_DELETE";
    GatewayEvent["ThreadListSync"] = "THREAD_LIST_SYNC";
    GatewayEvent["ThreadMemberUpdate"] = "THREAD_MEMBER_UPDATE";
    GatewayEvent["ThreadMembersUpdate"] = "THREAD_MEMBERS_UPDATE";
    GatewayEvent["GuildCreate"] = "GUILD_CREATE";
    GatewayEvent["GuildUpdate"] = "GUILD_UPDATE";
    GatewayEvent["GuildDelete"] = "GUILD_DELETE";
    GatewayEvent["GuildAuditLogEntryCreate"] = "GUILD_AUDIT_LOG_ENTRY_CREATE";
    GatewayEvent["GuildBanAdd"] = "GUILD_BAN_ADD";
    GatewayEvent["GuildBanRemove"] = "GUILD_BAN_REMOVE";
    GatewayEvent["GuildEmojisUpdate"] = "GUILD_EMOJIS_UPDATE";
    GatewayEvent["GuildStickersUpdate"] = "GUILD_STICKERS_UPDATE";
    GatewayEvent["GuildIntegrationsUpdate"] = "GUILD_INTEGRATIONS_UPDATE";
    GatewayEvent["GuildMemberAdd"] = "GUILD_MEMBER_ADD";
    GatewayEvent["GuildMemberRemove"] = "GUILD_MEMBER_REMOVE";
    GatewayEvent["GuildMemberUpdate"] = "GUILD_MEMBER_UPDATE";
    GatewayEvent["GuildMembersChunk"] = "GUILD_MEMBERS_CHUNK";
    GatewayEvent["GuildRoleCreate"] = "GUILD_ROLE_CREATE";
    GatewayEvent["GuildRoleUpdate"] = "GUILD_ROLE_UPDATE";
    GatewayEvent["GuildRoleDelete"] = "GUILD_ROLE_DELETE";
    GatewayEvent["GuildScheduledEventCreate"] = "GUILD_SCHEDULED_EVENT_CREATE";
    GatewayEvent["GuildScheduledEventUpdate"] = "GUILD_SCHEDULED_EVENT_UPDATE";
    GatewayEvent["GuildScheduledEventDelete"] = "GUILD_SCHEDULED_EVENT_DELETE";
    GatewayEvent["GuildScheduledEventUserAdd"] = "GUILD_SCHEDULED_EVENT_USER_ADD";
    GatewayEvent["GuildScheduledEventUserRemove"] = "GUILD_SCHEDULED_EVENT_USER_REMOVE";
    GatewayEvent["IntegrationCreate"] = "INTEGRATION_CREATE";
    GatewayEvent["IntegrationUpdate"] = "INTEGRATION_UPDATE";
    GatewayEvent["IntegrationDelete"] = "INTEGRATION_DELETE";
    GatewayEvent["InteractionCreate"] = "INTERACTION_CREATE";
    GatewayEvent["InviteCreate"] = "INVITE_CREATE";
    GatewayEvent["InviteDelete"] = "INVITE_DELETE";
    GatewayEvent["MessageCreate"] = "MESSAGE_CREATE";
    GatewayEvent["MessageUpdate"] = "MESSAGE_UPDATE";
    GatewayEvent["MessageDelete"] = "MESSAGE_DELETE";
    GatewayEvent["MessageDeleteBulk"] = "MESSAGE_DELETE_BULK";
    GatewayEvent["MessageReactionAdd"] = "MESSAGE_REACTION_ADD";
    GatewayEvent["MessageReactionRemove"] = "MESSAGE_REACTION_REMOVE";
    GatewayEvent["MessageReactionRemoveAll"] = "MESSAGE_REACTION_REMOVE_ALL";
    GatewayEvent["MessageReactionRemoveEmoji"] = "MESSAGE_REACTION_REMOVE_EMOJI";
    GatewayEvent["PresenceUpdate"] = "PRESENCE_UPDATE";
    GatewayEvent["StageInstanceCreate"] = "STAGE_INSTANCE_CREATE";
    GatewayEvent["StageInstanceUpdate"] = "STAGE_INSTANCE_UPDATE";
    GatewayEvent["StageInstanceDelete"] = "STAGE_INSTANCE_DELETE";
    GatewayEvent["TypingStart"] = "TYPING_START";
    GatewayEvent["UserUpdate"] = "USER_UPDATE";
    GatewayEvent["VoiceStateUpdate"] = "VOICE_STATE_UPDATE";
    GatewayEvent["VoiceServerUpdate"] = "VOICE_SERVER_UPDATE";
    GatewayEvent["WebhookUpdate"] = "WEBHOOK_UPDATE";
    GatewayEvent["MessagePollVoteAdd"] = "MESSAGE_POLL_VOTE_ADD";
    GatewayEvent["MessagePollVoteRemove"] = "MESSAGE_POLL_VOTE_REMOVE";
})(GatewayEvent || (GatewayEvent = {}));
export var GatewayOpCode;
(function (GatewayOpCode) {
    GatewayOpCode[GatewayOpCode["Dispatch"] = 0] = "Dispatch";
    GatewayOpCode[GatewayOpCode["Heartbeat"] = 1] = "Heartbeat";
    GatewayOpCode[GatewayOpCode["Identify"] = 2] = "Identify";
    GatewayOpCode[GatewayOpCode["PresenceUpdate"] = 3] = "PresenceUpdate";
    GatewayOpCode[GatewayOpCode["VoiceStateUpdate"] = 4] = "VoiceStateUpdate";
    GatewayOpCode[GatewayOpCode["Resume"] = 6] = "Resume";
    GatewayOpCode[GatewayOpCode["Reconnect"] = 7] = "Reconnect";
    GatewayOpCode[GatewayOpCode["RequestGuildMembers"] = 8] = "RequestGuildMembers";
    GatewayOpCode[GatewayOpCode["InvalidSession"] = 9] = "InvalidSession";
    GatewayOpCode[GatewayOpCode["Hello"] = 10] = "Hello";
    GatewayOpCode[GatewayOpCode["HeartbeatACK"] = 11] = "HeartbeatACK";
})(GatewayOpCode || (GatewayOpCode = {}));
export var Intents;
(function (Intents) {
    Intents[Intents["GUILDS"] = 1] = "GUILDS";
    Intents[Intents["GUILD_MEMBERS"] = 2] = "GUILD_MEMBERS";
    Intents[Intents["GUILD_MODERATION"] = 4] = "GUILD_MODERATION";
    Intents[Intents["GUILD_EMOJI_AND_STICKERS"] = 8] = "GUILD_EMOJI_AND_STICKERS";
    Intents[Intents["GUILD_INTEGRATIONS"] = 16] = "GUILD_INTEGRATIONS";
    Intents[Intents["GUILD_WEBHOOKS"] = 32] = "GUILD_WEBHOOKS";
    Intents[Intents["GUILD_INVITES"] = 64] = "GUILD_INVITES";
    Intents[Intents["GUILD_VOICE_STATES"] = 128] = "GUILD_VOICE_STATES";
    Intents[Intents["GUILD_PRESENCES"] = 256] = "GUILD_PRESENCES";
    Intents[Intents["GUILD_MESSAGES"] = 512] = "GUILD_MESSAGES";
    Intents[Intents["GUILD_MESSAGE_REACTIONS"] = 1024] = "GUILD_MESSAGE_REACTIONS";
    Intents[Intents["GUILD_MESSAGE_TYPING"] = 2048] = "GUILD_MESSAGE_TYPING";
    Intents[Intents["DIRECT_MESSAGES"] = 4096] = "DIRECT_MESSAGES";
    Intents[Intents["DIRECT_MESSAGE_REACTIONS"] = 8192] = "DIRECT_MESSAGE_REACTIONS";
    Intents[Intents["DIRECT_MESSAGE_TYPING"] = 16384] = "DIRECT_MESSAGE_TYPING";
    Intents[Intents["MESSAGE_CONTENT"] = 32768] = "MESSAGE_CONTENT";
    Intents[Intents["GUILD_SCHEDULED_EVENTS"] = 65536] = "GUILD_SCHEDULED_EVENTS";
    Intents[Intents["AUTO_MODERATION_CONFIGURATION"] = 1048576] = "AUTO_MODERATION_CONFIGURATION";
    Intents[Intents["AUTO_MODERATION_EXECUTION"] = 2097152] = "AUTO_MODERATION_EXECUTION";
    Intents[Intents["GUILD_MESSAGE_POLLS"] = 16777216] = "GUILD_MESSAGE_POLLS";
    Intents[Intents["DIRECT_MESSAGE_POLLS"] = 33554432] = "DIRECT_MESSAGE_POLLS";
})(Intents || (Intents = {}));
export var GatewayCloseEventCode;
(function (GatewayCloseEventCode) {
    GatewayCloseEventCode[GatewayCloseEventCode["UnknownError"] = 4000] = "UnknownError";
    GatewayCloseEventCode[GatewayCloseEventCode["UnknownOpCode"] = 4001] = "UnknownOpCode";
    GatewayCloseEventCode[GatewayCloseEventCode["DecodeError"] = 4002] = "DecodeError";
    GatewayCloseEventCode[GatewayCloseEventCode["NotAuthenticated"] = 4003] = "NotAuthenticated";
    GatewayCloseEventCode[GatewayCloseEventCode["AuthenticationFailed"] = 4004] = "AuthenticationFailed";
    GatewayCloseEventCode[GatewayCloseEventCode["AlreadyAuthenticated"] = 4005] = "AlreadyAuthenticated";
    GatewayCloseEventCode[GatewayCloseEventCode["InvalidSequenceNumber"] = 4007] = "InvalidSequenceNumber";
    GatewayCloseEventCode[GatewayCloseEventCode["RateLimited"] = 4008] = "RateLimited";
    GatewayCloseEventCode[GatewayCloseEventCode["SessionTimedOut"] = 4009] = "SessionTimedOut";
    GatewayCloseEventCode[GatewayCloseEventCode["InvalidShard"] = 4010] = "InvalidShard";
    GatewayCloseEventCode[GatewayCloseEventCode["ShardingRequired"] = 4011] = "ShardingRequired";
    GatewayCloseEventCode[GatewayCloseEventCode["InvalidAPIVersion"] = 4012] = "InvalidAPIVersion";
    GatewayCloseEventCode[GatewayCloseEventCode["InvalidIntents"] = 4013] = "InvalidIntents";
    GatewayCloseEventCode[GatewayCloseEventCode["DisallowedIntents"] = 4014] = "DisallowedIntents";
})(GatewayCloseEventCode || (GatewayCloseEventCode = {}));
