// lilybird/dist/enums/guild.js
export var VerificationLevel;
(function (VerificationLevel) {
    VerificationLevel[VerificationLevel["NONE"] = 0] = "NONE";
    VerificationLevel[VerificationLevel["LOW"] = 1] = "LOW";
    VerificationLevel[VerificationLevel["MEDIUM"] = 2] = "MEDIUM";
    VerificationLevel[VerificationLevel["HIGH"] = 3] = "HIGH";
    VerificationLevel[VerificationLevel["VERY_HIGH"] = 4] = "VERY_HIGH";
})(VerificationLevel || (VerificationLevel = {}));
export var DefaultMessageNotificationLevel;
(function (DefaultMessageNotificationLevel) {
    DefaultMessageNotificationLevel[DefaultMessageNotificationLevel["ALL_MESSAGES"] = 0] = "ALL_MESSAGES";
    DefaultMessageNotificationLevel[DefaultMessageNotificationLevel["ONLY_MENTIONS"] = 1] = "ONLY_MENTIONS";
})(DefaultMessageNotificationLevel || (DefaultMessageNotificationLevel = {}));
export var ExplicitContentFilterLevel;
(function (ExplicitContentFilterLevel) {
    ExplicitContentFilterLevel[ExplicitContentFilterLevel["DISABLED"] = 0] = "DISABLED";
    ExplicitContentFilterLevel[ExplicitContentFilterLevel["MEMBERS_WITHOUT_ROLES"] = 1] = "MEMBERS_WITHOUT_ROLES";
    ExplicitContentFilterLevel[ExplicitContentFilterLevel["ALL_MEMBERS"] = 2] = "ALL_MEMBERS";
})(ExplicitContentFilterLevel || (ExplicitContentFilterLevel = {}));
export var MFALevel;
(function (MFALevel) {
    MFALevel[MFALevel["NONE"] = 0] = "NONE";
    MFALevel[MFALevel["ELEVATED"] = 1] = "ELEVATED";
})(MFALevel || (MFALevel = {}));
export var SystemChannelFlags;
(function (SystemChannelFlags) {
    SystemChannelFlags[SystemChannelFlags["SUPPRESS_JOIN_NOTIFICATIONS"] = 1] = "SUPPRESS_JOIN_NOTIFICATIONS";
    SystemChannelFlags[SystemChannelFlags["SUPPRESS_PREMIUM_SUBSCRIPTIONS"] = 2] = "SUPPRESS_PREMIUM_SUBSCRIPTIONS";
    SystemChannelFlags[SystemChannelFlags["SUPPRESS_GUILD_REMINDER_NOTIFICATIONS"] = 4] = "SUPPRESS_GUILD_REMINDER_NOTIFICATIONS";
    SystemChannelFlags[SystemChannelFlags["SUPPRESS_JOIN_NOTIFICATION_REPLIES"] = 8] = "SUPPRESS_JOIN_NOTIFICATION_REPLIES";
    SystemChannelFlags[SystemChannelFlags["SUPPRESS_ROLE_SUBSCRIPTION_PURCHASE_NOTIFICATIONS"] = 16] = "SUPPRESS_ROLE_SUBSCRIPTION_PURCHASE_NOTIFICATIONS";
    SystemChannelFlags[SystemChannelFlags["SUPPRESS_ROLE_SUBSCRIPTION_PURCHASE_NOTIFICATION_REPLIES"] = 32] = "SUPPRESS_ROLE_SUBSCRIPTION_PURCHASE_NOTIFICATION_REPLIES";
})(SystemChannelFlags || (SystemChannelFlags = {}));
export var PremiumTier;
(function (PremiumTier) {
    PremiumTier[PremiumTier["NONE"] = 0] = "NONE";
    PremiumTier[PremiumTier["TIER_1"] = 1] = "TIER_1";
    PremiumTier[PremiumTier["TIER_2"] = 2] = "TIER_2";
    PremiumTier[PremiumTier["TIER_3"] = 3] = "TIER_3";
})(PremiumTier || (PremiumTier = {}));
export var GuildNSFWLevel;
(function (GuildNSFWLevel) {
    GuildNSFWLevel[GuildNSFWLevel["DEFAULT"] = 0] = "DEFAULT";
    GuildNSFWLevel[GuildNSFWLevel["EXPLICIT"] = 1] = "EXPLICIT";
    GuildNSFWLevel[GuildNSFWLevel["SAFE"] = 2] = "SAFE";
    GuildNSFWLevel[GuildNSFWLevel["AGE_RESTRICTED"] = 3] = "AGE_RESTRICTED";
})(GuildNSFWLevel || (GuildNSFWLevel = {}));
export var GuildMemberFlag;
(function (GuildMemberFlag) {
    GuildMemberFlag[GuildMemberFlag["DID_REJOIN"] = 1] = "DID_REJOIN";
    GuildMemberFlag[GuildMemberFlag["COMPLETED_ONBOARDING"] = 2] = "COMPLETED_ONBOARDING";
    GuildMemberFlag[GuildMemberFlag["BYPASSES_VERIFICATION"] = 4] = "BYPASSES_VERIFICATION";
    GuildMemberFlag[GuildMemberFlag["STARTED_ONBOARDING"] = 8] = "STARTED_ONBOARDING";
    GuildMemberFlag[GuildMemberFlag["IS_GUEST"] = 16] = "IS_GUEST";
    GuildMemberFlag[GuildMemberFlag["STARTED_HOME_ACTIONS"] = 32] = "STARTED_HOME_ACTIONS";
    GuildMemberFlag[GuildMemberFlag["COMPLETED_HOME_ACTIONS"] = 64] = "COMPLETED_HOME_ACTIONS";
    GuildMemberFlag[GuildMemberFlag["AUTOMOD_QUARANTINED_USERNAME"] = 128] = "AUTOMOD_QUARANTINED_USERNAME";
    GuildMemberFlag[GuildMemberFlag["DM_SETTINGS_UPSELL_ACKNOWLEDGED"] = 256] = "DM_SETTINGS_UPSELL_ACKNOWLEDGED";
})(GuildMemberFlag || (GuildMemberFlag = {}));
export var IntegrationType;
(function (IntegrationType) {
    IntegrationType["Twitch"] = "twitch";
    IntegrationType["Youtube"] = "youtube";
    IntegrationType["Discord"] = "discord";
    IntegrationType["GuildSubscription"] = "guild_subscription";
})(IntegrationType || (IntegrationType = {}));
export var IntegrationExpireBehavior;
(function (IntegrationExpireBehavior) {
    IntegrationExpireBehavior[IntegrationExpireBehavior["RemoveRole"] = 0] = "RemoveRole";
    IntegrationExpireBehavior[IntegrationExpireBehavior["Role"] = 1] = "Role";
})(IntegrationExpireBehavior || (IntegrationExpireBehavior = {}));
export var InviteTargetType;
(function (InviteTargetType) {
    InviteTargetType[InviteTargetType["STREAM"] = 1] = "STREAM";
    InviteTargetType[InviteTargetType["EMBEDDED_APPLICATION"] = 2] = "EMBEDDED_APPLICATION";
})(InviteTargetType || (InviteTargetType = {}));
export var RoleFlags;
(function (RoleFlags) {
    RoleFlags[RoleFlags["IN_PROMPT"] = 1] = "IN_PROMPT";
})(RoleFlags || (RoleFlags = {}));
export var PrivacyLevel;
(function (PrivacyLevel) {
    PrivacyLevel[PrivacyLevel["PUBLIC"] = 1] = "PUBLIC";
    PrivacyLevel[PrivacyLevel["GUILD_ONLY"] = 2] = "GUILD_ONLY";
})(PrivacyLevel || (PrivacyLevel = {}));
export var OnboardingMode;
(function (OnboardingMode) {
    OnboardingMode[OnboardingMode["ONBOARDING_DEFAULT"] = 0] = "ONBOARDING_DEFAULT";
    OnboardingMode[OnboardingMode["ONBOARDING_ADVANCED"] = 1] = "ONBOARDING_ADVANCED";
})(OnboardingMode || (OnboardingMode = {}));
export var PromptType;
(function (PromptType) {
    PromptType[PromptType["MULTIPLE_CHOICE"] = 0] = "MULTIPLE_CHOICE";
    PromptType[PromptType["DROPDOWN"] = 1] = "DROPDOWN";
})(PromptType || (PromptType = {}));
