// lilybird/dist/enums/others.js
export var Locale;
(function (Locale) {
    Locale["Indonesian"] = "id";
    Locale["Danish"] = "da";
    Locale["German"] = "de";
    Locale["EnglishUK"] = "en-GB";
    Locale["EnglishUS"] = "en-US";
    Locale["Spanish"] = "es-ES";
    Locale["French"] = "fr";
    Locale["Croatian"] = "hr";
    Locale["Italian"] = "it";
    Locale["Lithuanian"] = "lt";
    Locale["Hungarian"] = "hu";
    Locale["Dutch"] = "nl";
    Locale["Norwegian"] = "no";
    Locale["Polish"] = "pl";
    Locale["PortugueseBrazilian"] = "pt-BR";
    Locale["Romanian"] = "ro";
    Locale["Finnish"] = "fi";
    Locale["Swedish"] = "sv-SE";
    Locale["Vietnamese"] = "vi";
    Locale["Turkish"] = "tr";
    Locale["Czech"] = "cs";
    Locale["Greek"] = "el";
    Locale["Bulgarian"] = "bg";
    Locale["Russian"] = "ru";
    Locale["Ukrainian"] = "uk";
    Locale["Hindi"] = "hi";
    Locale["Thai"] = "th";
    Locale["ChineseChina"] = "zh-CN";
    Locale["Japanese"] = "ja";
    Locale["ChineseTaiwan"] = "zh-TW";
    Locale["Korean"] = "ko";
})(Locale || (Locale = {}));
export var MembershipState;
(function (MembershipState) {
    MembershipState[MembershipState["INVITED"] = 1] = "INVITED";
    MembershipState[MembershipState["ACCEPTED"] = 2] = "ACCEPTED";
})(MembershipState || (MembershipState = {}));
